velox/docs/develop/hash-table.rst

```diff
@@ -0,0 +1,147 @@
+Remember that capacity is always a power of 2: :m:`2^n`. Each bucket stored 16 entries, hence, we
+need :m:`2^{(n-4)}` buckets to store :m:`2^n` entries. Hence, we need to use n-4 bits of the hash for the bucket.
+
+Let's say we have a hash table that can store a million entries: :m:`2^{20}` = 1,048,576. Here, n = 20 and
+N = n - 4 = 16. We will use 16 bits for the bucket.
+
+Given hash number:
+
+01011011 11001010 01111100 0\ :raw-html:`<text style="background-color:orange">1101001</text>` 10110111
+1\ :raw-html:`<text style="background-color:#ADD8E6">0010100 11111000 1</text>`\ 1001110
+
+We compute the tag as 1\ :raw-html:`<text style="background-color:orange">1101001</text>` and bucket offset
+as 1,374,336 (00000000 00000000 00000000 00000000 00000000
+0\ :raw-html:`<text style="background-color:#ADD8E6">0010100 11111000 1</text>`\ 0000000).
+Bucket offset is the number of bytes from the start of the hash table.
+
+.. code-block:: c++
+
+    bucket offset = bucket number * 128
+    bucket number = bucket offset / 128
+
+The bucket offset is used to get to the bucket, in this case bucket number 10737.
+The candidate bucket can be empty, partially filled or full.
+
+**The bucket is empty.**
+
+.. image:: images/ht-empty.png
+    :align: center
+
+In this case, we simply insert our entry into the first slot of the bucket.
+We store only the tag (7 bits of the hash value) in the hash table. The hash number itself is not stored.
+The pointer refers to memory outside of the hash table where the complete value is stored.
+This is usually a row in a RowContainer. The hash table can be viewed as an index on top of a RowContainer
+that helps locate records with matching keys faster. The hash table itself doesn’t store data or keys.
+
+.. image:: images/ht-insert1.png
+    :align: center
+
+**The bucket is partially filled.**
+
+For example, the bucket has one slot occupied (like above).
+In this case there is a possibility that the new entry is a duplicate of the one already stored.
+Hence, we compare the tag of the new entry with tags stored in the bucket.
+If none match, this entry is not a duplicate, so we store it in the next available slot in the bucket.
+
+.. image:: images/ht-insert2.png
+    :align: center
+
+However, if one or more existing tags match the tag of the new entry, we follow the pointers to compare
+the keys to determine whether there is a match or not. If there is no match we insert a new entry.
+Otherwise, there is a duplicate. The row is chained to the row list pointed to by the row entry
+and no new entry is inserted.
+
+**The bucket is full.**
+
+.. image:: images/ht-full.png
+    :align: center
+
+First, we need to check if the new entry is a duplicate of one of the 16 entries stored in the bucket.
+We compare tags and if necessary follow the pointers to compare the keys.
+If there is a match, we drop the entry and we are done. If there is no match, we go to the
```

Comment:
> we drop the entry and we are done

nit: perhaps, match the above

"The row is chained to the row list pointed to by the row entry
and no new entry is inserted."